runtime-modules/creator-token/crt-core/src/types.rs

```diff
@@ -0,0 +1,184 @@
+use codec::{Decode, Encode};
+use frame_support::{dispatch::DispatchResult, ensure};
+use sp_arithmetic::traits::{Saturating, Zero};
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq)]
+pub struct AccountData<Balance> {
+    /// Non-reserved part of the balance. There may still be restrictions
+    /// on this, but it is the total pool what may in principle be
+    /// transferred, reserved and used for tipping.
+    free_balance: Balance,
+
+    /// This balance is a 'reserve' balance that other subsystems use
+    /// in order to set aside tokens that are still 'owned' by the
+    /// account holder, but which are not usable in any case.
+    frozen_balance: Balance,
+}
+#[derive(Encode, Decode, Clone, Copy, PartialEq, Eq)]
+pub enum MaxTotalIssuance<Balance> {
+    Unlimited,
+    Limited(Balance),
+}
+
+impl<Balance> Default for MaxTotalIssuance<Balance> {
+    fn default() -> Self {
+        MaxTotalIssuance::<Balance>::Unlimited
+    }
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Default)]
+pub struct TokenData<Balance, Hash> {
+    /// Current token issuanc
+    current_total_issuance: Balance,
+
+    /// Max total issuance allowed for the token
+    max_total_issuance: MaxTotalIssuance<Balance>,
+
+    /// Hash of a human-readable description
+    description: Hash,
+
+    /// Existential deposit allowed for the token
+    existential_deposit: Balance,
+}
+
+/// Default trait for AccountData
+impl<Balance: Zero> Default for AccountData<Balance> {
+    fn default() -> Self {
+        Self {
+            free_balance: Balance::zero(),
+            frozen_balance: Balance::zero(),
+        }
+    }
+}
+
+/// Interface for interacting with AccountData
+impl<Balance: Copy + PartialOrd + Saturating + Zero> AccountData<Balance> {
+    /// Retrieve free balance amount
+    pub(crate) fn free_balance(&self) -> Balance {
+        self.free_balance
+    }
+
+    /// Retrieve frozen balance amount
+    pub(crate) fn frozen_balance(&self) -> Balance {
+        self.frozen_balance
+    }
+
+    /// checks whether free balance has enough amount to freeze
+    pub(crate) fn can_freeze<T: crate::Trait>(&self, amount: Balance) -> DispatchResult {
+        ensure!(
+            self.free_balance >= amount,
+            crate::Error::<T>::InsufficientFreeBalanceForFreezing,
+        );
+        Ok(())
+    }
+
+    /// freeze specified amount: infallible
+    pub(crate) fn freeze(&mut self, amount: Balance) {
```

Comment:
Again, if we are talking about staking here, lets call this `reserve`, consistently.